quantinuumqircheck/qircheck.py

```diff
@@ -0,0 +1,320 @@
+
+    functions: list[Function]
+    A list of functions (declared or defined) in pyqir's representation of a QIR program.
+
+    Returns
+    -------
+    bool"""
+    if isinstance(instr, pq.Call):
+        return is_wasm_call(instr, functions)
+    return False
+
+
+def is_wasm_call(fun_call: pq.Call, functions: list[pq.Function]) -> bool:
+    """Indicate whether a QIR function call is a call to a wasm function
+
+    Parameters
+    ----------
+    fun_call: Call
+    A function call instruction in pyqir's representation of a QIR program.
+
+    functions: list[Function]
+    A list of functions (declared or defined) in pyqir's representation of a QIR program.
+
+    Returns
+    -------
+    bool"""
+    return has_wasm_attr(fun_call.callee.name, functions)
+
+
+def has_wasm_attr(gate_name: str, functions: list[pq.Function]) -> bool:
+    """Indicates whether the QIR gate's function definition was marked with the { "wasm" } group attribute
+
+    Parameters
+    ----------
+    gate_name: str
+    represents the name of the gate_name that we are inspecting the attributes for
+
+    functions: list[Function]
+    a list of pyqir function objects corresponding to QIR function declarations/definitions
+    These can have attributes defined on them
+
+    Returns
+    -------
+    bool"""
+    for fun in functions:
+        if fun.name == gate_name:
+            return "wasm" in fun.attributes.func
+    return False
+
+
+def is_classical_op(instr: pq.Instruction) -> bool:
+    op_opcodes = [
+        pq.Opcode.ADD,
+        pq.Opcode.SUB,
+        pq.Opcode.MUL,
+        pq.Opcode.UDIV,
+        pq.Opcode.SDIV,
+        pq.Opcode.UREM,
+        pq.Opcode.SREM,
+        pq.Opcode.AND,
+        pq.Opcode.OR,
+        pq.Opcode.XOR,
+        pq.Opcode.SHL,
+        pq.Opcode.LSHR,
+        pq.Opcode.ASHR,
+        pq.Opcode.FADD,
+        pq.Opcode.FSUB,
+        pq.Opcode.FMUL,
+        pq.Opcode.FDIV,
+        pq.Opcode.FREM,
+        pq.Opcode.FNEG,
+        pq.Opcode.ICMP,
+        pq.Opcode.FCMP,
+        pq.Opcode.ZEXT,
+        pq.Opcode.SELECT,
+    ]
+    return isinstance(instr, pq.Instruction) and instr.opcode in op_opcodes
+
+
+class ValidationError:
+    # pylint: disable=too-few-public-methods
+    def __init__(self, instr: pq.Instruction, line_num: int):
+        self.error_message = f"Encountered Unexpected Instruction {str(instr)}"
+        self.line_num = line_num
+
+
+def validate_qir_base(qir_prog: pq.Module) -> Union[bool, ValidationError]:
+    """Validate that the QIR corresponds to a valid quantinuum profile
+    program that we can support, allowing pytket creg functions and"""
+
+    def _is_valid_call_help(instr: pq.Instruction) -> bool:
+        return is_valid_call(instr, qir_prog.functions)
+
+    if not isinstance(qir_prog.functions, list):
+        raise ValueError(
+            "Expected the QIR file to have at least one function but none was found",
+            -1,
```

Comment:
Does the -1 have any significance (here and elsewhere)?

Reply:
done in [remove -1 from error](https://github.com/CQCL/quantinuum-qircheck/pull/1/commits/002866bf9e7bdb40d851d5535eca98ff4df0e5e7)
[002866b](https://github.com/CQCL/quantinuum-qircheck/pull/1/commits/002866bf9e7bdb40d851d5535eca98ff4df0e5e7)